gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/converter/EnvelopeSchemaConverter.java

```diff
@@ -1,152 +1,145 @@
- * (the "License"); you may not use this file except in compliance with
- * the License.  You may obtain a copy of the License at
- *
- *    http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
 package org.apache.gobblin.converter;
 
-import com.google.common.base.Optional;
-import com.google.common.cache.CacheBuilder;
-import com.google.common.cache.CacheLoader;
-import com.google.common.cache.LoadingCache;
-import org.apache.gobblin.configuration.ConfigurationKeys;
-import org.apache.gobblin.configuration.WorkUnitState;
-import org.apache.gobblin.converter.filter.AvroProjectionConverter;
-import org.apache.gobblin.converter.filter.AvroSchemaFieldRemover;
-import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistry;
-import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistryFactory;
-import org.apache.gobblin.metrics.kafka.SchemaRegistryException;
-import org.apache.gobblin.util.AvroUtils;
-import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
-import java.util.concurrent.ExecutionException;
-import javax.xml.bind.DatatypeConverter;
+
 import org.apache.avro.Schema;
 import org.apache.avro.generic.GenericDatumReader;
 import org.apache.avro.generic.GenericRecord;
 import org.apache.avro.io.Decoder;
 import org.apache.avro.io.DecoderFactory;
+import org.apache.gobblin.configuration.WorkUnitState;
+import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistry;
+import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistryFactory;
+import org.apache.gobblin.util.AvroUtils;
+
+import com.google.common.base.Optional;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import javax.xml.bind.DatatypeConverter;
+
 
 /**
- * A converter for extracting schema/records from an envelope schema.
- * Input schema: envelope schema - must have fields payloadSchemaId (the schema registry key of the output
- *               schema) and payload (byte data for output record)
- * Input record: record corresponding to input schema
- * Output schema: schema obtained from schema registry using key provided in input record's {@link #PAYLOAD_SCHEMA_ID_FIELD}
- * Output record: record corresponding to output schema obtained from input record's {@link #PAYLOAD_FIELD} as bytes
+ * Base class for an envelope schema converter using {@link KafkaSchemaRegistry}
  */
-public class EnvelopeSchemaConverter extends Converter<Schema, String, GenericRecord, GenericRecord> {
+public abstract class EnvelopeSchemaConverter<P> extends Converter<Schema, Schema, GenericRecord, GenericRecord> {
```

Comment:
What about backwards compatibility?

Reply:
I guess the converter is rarely used because it returns `String` as the output schema. The cost for this backwards incompatible change would be 0 or negligible.